packages/patterns/integration/shadow-dom-utils.ts

```diff
@@ -0,0 +1,170 @@
+/**
+ * Utility functions for navigating through shadow DOM boundaries in tests
+ */
+
+type Page = any; // The Page type from integration utils
+
+/**
+ * Traverses through shadow DOM boundaries following a path of selectors
+ * @param page - The page object
+ * @param path - Array of selectors to traverse through shadow DOMs
+ * @returns The element at the end of the path, or null if not found
+ */
+export async function traverseShadowDOM(
+  page: Page,
+  path: string[],
+): Promise<any> {
```

Comment:
What do `Element`s look like converted over the boundary? I don't recall if Astral translates it directly to an `ElementHandle`, but that'd be very helpful if it did -- is this function used?